### PR TITLE
AUTH-2235: Validate XML in SAML

### PR DESCRIPTION
**Pull request by SruthiD-Invision**
Added XML Validation.

Summary:

- Used https://github.com/mattermost/xml-roundtrip-validator and validated XML
- Fixed and added unit tests

### CODE REVIEW

File: service_provider.go [hunk 1]
```diff
@@ -390,12 +392,19 @@ func (sp *ServiceProvider) ParseResponse(req *http.Request, possibleRequestIDs [
 	}
 	retErr.Response = string(rawResponseBuf)
 
+	// XML Validation
+	if err := xrv.Validate(strings.NewReader(string(rawResponseBuf))); err != nil {
+		retErr.PrivateErr = fmt.Errorf("invalid xml response: %s", err)
```
Comment (mikehepple) [inline service_provider.go:397]: should this be `%w`? (honest question, I don't know norms for this codebase)

File: service_provider.go [hunk 2]
```diff
@@ -390,12 +392,19 @@ func (sp *ServiceProvider) ParseResponse(req *http.Request, possibleRequestIDs [
...
+		return nil, requestID, retErr
+	}
+
 	// do some validation first before we decrypt
 	resp := Response{}
 	if err := xml.Unmarshal(rawResponseBuf, &resp); err != nil {
-		retErr.PrivateErr = fmt.Errorf("cannot unmarshal response: %s", err)
+		retErr.PrivateErr = fmt.Errorf("cannot unmarshal response: %v", err)
```
Comment (mikehepple) [inline service_provider.go:404]: as above

**Review by mikehepple — APPROVED** [review]
Couple of non-essential questions on error wrapping.

Bigger question - whats our test plan to assure that our SAML producers are sending us valid XML?

File: service_provider.go [hunk 3]
```diff
@@ -390,12 +392,19 @@ func (sp *ServiceProvider) ParseResponse(req *http.Request, possibleRequestIDs [
 	}
 	retErr.Response = string(rawResponseBuf)
 
+	// XML Validation
```
Comment (mikehepple) [inline service_provider.go:395]: Might it be worth linking out to some details on the CVE? Like the blog post?

**Review by srfrog — APPROVED** [review]

**Pull request merged**